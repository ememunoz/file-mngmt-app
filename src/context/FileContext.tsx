import AsyncStorage from '@react-native-async-storage/async-storage'
import {
  ReactNode,
  createContext,
  useContext,
  useEffect,
  useState,
} from 'react'

type ImageInfoType = {
  src: string
}

type SheetInfoType = {
  src: string
  total: number
  name: string
}

type FileStorageType = {
  imageInfoArray: ImageInfoType[]
  sheetInfoArray: SheetInfoType[]
}

type FileContextType = {
  imageInfoArray: ImageInfoType[]
  sheetInfoArray: SheetInfoType[]
  addImage: (src: string) => void
  addSheet: (obj: SheetInfoType) => void
}

const FileContext = createContext<FileContextType>({
  imageInfoArray: [],
  sheetInfoArray: [],
  addImage: () => {},
  addSheet: () => {},
})

const getInitialFileContext = async () => {
  const defaultValue: FileStorageType = {
    imageInfoArray: [],
    sheetInfoArray: [],
  }
  try {
    const rawValue = await AsyncStorage.getItem('@file-mngmt-state')
    const jsonValue: FileStorageType = !!rawValue
      ? (JSON.parse(rawValue) as FileStorageType)
      : defaultValue
    return jsonValue
  } catch (e) {
    return defaultValue
  }
}

export const FileContextProvider = ({ children }: { children: ReactNode }) => {
  const [imageInfoArray, setImageInfoArray] = useState<ImageInfoType[]>([])
  const [sheetInfoArray, setSheetInfoArray] = useState<SheetInfoType[]>([])

  useEffect(() => {
    const getDefaultValues = async () => {
      const { imageInfoArray, sheetInfoArray } = await getInitialFileContext()
      setImageInfoArray(imageInfoArray)
      setSheetInfoArray(sheetInfoArray)
    }
    getDefaultValues()
  }, [])

  const updateStorage = async () => {
    const data: FileStorageType = {
      imageInfoArray,
      sheetInfoArray,
    }
    await AsyncStorage.setItem('@file-mngmt-state', JSON.stringify(data))
  }

  const addImage = (src: string) => {
    const newArray = [...imageInfoArray, { src }]
    setImageInfoArray(newArray)
    updateStorage()
  }

  const addSheet = ({ src, name, total }: SheetInfoType) => {
    const newArray = [...sheetInfoArray, { src, name, total }]
    setSheetInfoArray(newArray)
    updateStorage()
  }

  return (
    <FileContext.Provider
      value={{ imageInfoArray, sheetInfoArray, addImage, addSheet }}
    >
      {children}
    </FileContext.Provider>
  )
}

export const useFileContext = () => useContext(FileContext)
