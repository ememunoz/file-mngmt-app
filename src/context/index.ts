export { FileContextProvider, useFileContext } from './FileContext/FileContext'
