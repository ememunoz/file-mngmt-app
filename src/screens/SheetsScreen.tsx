import React from 'react'
import { FlatList, StyleSheet, Text, View } from 'react-native'

import Icon from '../components/Icon'
import { useFileContext } from '../context/FileContext'

type Props = {}

const SheetsScreen = () => {
  const { sheetInfoArray } = useFileContext()

  const renderItem = ({ item }) => {
    return (
      <View style={styles.itemContainer}>
        <View style={styles.imgContainer}>
          <Icon name="csv-file" height={64} width={64} fill="hsl(0, 0%, 0%)" />
        </View>
        <View style={styles.dataContainer}>
          <Text style={styles.dataFileName}>File name: {item.name}</Text>
          <Text style={styles.dataTotal}>Total: {item.total}</Text>
        </View>
      </View>
    )
  }

  return (
    <View style={styles.container}>
      {sheetInfoArray.length > 0 ? (
        <FlatList
          data={sheetInfoArray}
          keyExtractor={item => item.src}
          renderItem={renderItem}
        />
      ) : (
        <Text style={styles.message}>
          You haven't uploaded any sheets, you can upload them in home page! ðŸ˜„
        </Text>
      )}
    </View>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    margin: 16,
    justifyContent: 'center',
  },
  message: {
    alignSelf: 'center',
    fontSize: 24,
    color: 'hsl(0, 0%, 60%)',
    textAlign: 'center',
    marginHorizontal: 16,
  },

  itemContainer: {
    padding: 16,
    marginBottom: 16,
    flexDirection: 'row',
    backgroundColor: 'hsl(0, 100%, 100%)',
    shadowColor: 'hsl(0, 0%, 0%)',
    shadowOffset: {
      width: 2,
      height: 2,
    },
    shadowRadius: 2.6,
    shadowOpacity: 0.23,
    elevation: 4,
  },
  imgContainer: {
    flex: 1 / 4,
    alignItems: 'center',
    justifyContent: 'center',
  },
  dataContainer: {
    flex: 3 / 4,
  },
  dataFileName: {
    marginBottom: 8,
    fontWeight: 'bold',
  },
  dataTotal: {},
})

export default SheetsScreen
