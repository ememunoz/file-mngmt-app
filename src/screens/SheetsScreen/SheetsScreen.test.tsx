import { render, screen } from '@testing-library/react-native'
import React from 'react'

import { FileContextProvider } from '../../context'
import SheetsScreen from './SheetsScreen'

describe('<SheetsScreen />', () => {
  const component = (
    <FileContextProvider>
      <SheetsScreen />
    </FileContextProvider>
  )
  it('renders correctly', () => {
    render(component)
    expect(screen.toJSON()).toMatchSnapshot()
  })
})
