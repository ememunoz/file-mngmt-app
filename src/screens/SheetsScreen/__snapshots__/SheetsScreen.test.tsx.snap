// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SheetsScreen /> renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
      "margin": 16,
    }
  }
>
  <Text
    style={
      Object {
        "alignSelf": "center",
        "color": "hsl(0, 0%, 60%)",
        "fontSize": 24,
        "marginHorizontal": 16,
        "textAlign": "center",
      }
    }
  >
    You haven't uploaded any sheets, you can upload them in home page! ðŸ˜„
  </Text>
</View>
`;
